### ISSUE TITLE
Forward return type of injected dependencies with Depends()

### ISSUE THREAD

**Issue by tasercake** [tiangolo/fastapi#4750]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
"""
I have a function that specifies a return type, but that type is erased 
when I inject it into an endpoint using `Depends()`
"""
from fastapi import FastAPI, Depends

app = FastAPI()

async def dependency() -> int:
    return 0

@app.get("/")
async def root(dep = Depends(dependency)):
    # Get mypy to emit the return type of the value returned by the dependency
    reveal_type(dep)  # Revealed type is "Any"
```

### Description

When injecting `dependency` it into my view function with `Depends(dependency)`, the return type (`int`) is erased and replaced with `Any`.

This can be verified by running `mypy` against my code sample above (`reveal_type()` should cause mypy to emit the type it infers)

### Wanted Solution

It would be great if the return type of `dependency` were preserved, i.e.

```python
@app.get("/")
async def root(dep = Depends(dependency)):
    reveal_type(dep)  # Revealed type is "int"
```

---

I see in the definition of [Depends](https://github.com/tiangolo/fastapi/blob/eddbae948f04e13fe412dc45a569d10e34b698a4/fastapi/param_functions.py#L278-L281) that the signature is as follows:

```python
def Depends(dependency: Optional[Callable[..., Any]] = None, *, use_cache: bool = True) -> Any:
```

I'm no expert on the Python type system or Mypy, but I think the following could get Mypy to correctly infer the return type of `dependency`:

```python
from typing import TypeVar, Awaitable, Union

T_ = TypeVar("T_")
MaybeAwaitable = Union[T_, Awaitable[T_]]

T = TypeVar("T")
def Depends(dependency: Optional[Callable[..., MaybeAwaitable[T]]] = None, *, use_cache: bool = True) -> T:
    return params.Depends(dependency=dependency, use_cache=use_cache)
```

### Wanted Code

```python
"""
The return type of my dependency should be preserved 
when injecting it with `Depends(dependency)`
"""
from fastapi import FastAPI, Depends

app = FastAPI()

async def dependency() -> int:
    return 0

@app.get("/")
async def root(dep = Depends(dependency)):
    reveal_type(dep)  # Revealed type is "int"
```

### Operating System

macOS

### FastAPI Version

0.75.0

### Python Version

3.7.9

### Additional Context

My current approach is to explicitly annotate my injected dependencies, but this is prone to error since Mypy doesn't help catch these issues.

Consider:

```python
from fastapi import FastAPI, Depends

app = FastAPI()

# Returns a different type now
async def dict_dependency() -> dict:
    return {}

@app.get("/")
async def root(dict_dep: int = Depends(dict_dependency)):
	# But Mypy doesn't catch the issue
    reveal_type(dict_dep)  # Revealed type is "int"
```

**Comment by jd-0001** [comment]
FastAPI doesn't forward the type (I guess it's not possible) so you have to define the type explicitly just like your current approach.

Please refer to the docs: https://fastapi.tiangolo.com/tutorial/dependencies/#declare-the-dependency-in-the-dependant

**Comment by tasercake** [comment]
@jd-0001 You're right about that, but I was wondering if it might be possible to change that.

This feature could have saved me a lot of trouble when I made mistakes in explicitly annotating injected dependencies.

**Comment by odiseo0** [comment]
I honestly don't know if your use case if more complex than the one you posted but you could do the classic:
```py
from fastapi import FastAPI, Depends

app = FastAPI()

async def dependency() -> int:
    return 0

@app.get("/")
async def root(dep: int = Depends(dependency)):
    reveal_type(dep)
```

**Comment by tasercake** [comment]
@odiseo0 Yup that's my workaround for now, but see my additional comment about the issues that arise when the return type of the dependency diverges from the type annotation in the view function. I think in an ideal world mypy would catch discrepancies like this, but it doesn't seem possible right now since the type information is erased.

**Comment by odiseo0** [comment]
Understood, I don't know if you're in FastAPI Discord server but long ago there was a thread talking about Depends return type, I'll refer you to that thread: https://discord.com/channels/772607987558252564/921809701006610484

If you're not, this is the code that @Doctor wrote:
```py
import datetime
import typing
from typing import Callable, TypeVar, Any, Union, Generator, Optional

T = TypeVar("T")
DependencyFunction = Union[
    Callable[..., T],
    Generator[T, Any, Any],
    # Async callables and async generators/iterables could also be added here
]


@typing.overload
def Depends(dependency: None = None) -> Any:
    ...


@typing.overload
def Depends(dependency: DependencyFunction[T]) -> T:
    ...


def Depends(
    dependency: Optional[DependencyFunction[T]] = None,
) -> T:
    pass


def test_callable() -> int:
    ...


a = Depends(test_callable)
reveal_type(a)  # int

b = Depends()
reveal_type(b)  # Any

c: datetime.date = Depends()
reveal_type(c)  # datetime.date

# error: Incompatible types in assignment (expression has type "int", variable has type "date")
incompatible_type: datetime.date = Depends(test_callable)
```
Maybe we can create a PR if we manage to resolve the problem in a nice way

@Kludex What are your thoughts on this?

**Comment by Kludex** [comment]
Ideally FastAPI would use [`Annotated`](https://docs.python.org/3/library/typing.html#typing.Annotated) here.

**Comment by ThirVondukr** [comment]
@Kludex  I don't think type validation could be performed with mypy if you use `Annotated`:

```py
dependency = Annotated[AsyncSession, Depends(get_session)]
```
Though in this case type checking could be performed programmatically based on function return type

**Comment by Kludex** [comment]
Ah! Yep, mb. The above looks reasonable to me.

**Comment by ThirVondukr** [comment]
I'm just iterating over possible solutions, you mentioned `Annotated`, but you'd have to perform programmatic type checking/validation.
Using `@typing.overload` seems like a good solution to me, iirc you have to handle 3 cases:
- Implicit class dependencies, e.g. `service: Service = Depends()`
- Explicit class dependencies - `service = Depends(Service)`
- Functions and generators
All should be possible with base function that would return `Any` and overloads for `Type` and `Callable`s

**Comment by ThirVondukr** [comment]
I think something like this should do
I had to create an overload for simple callables, seems like i can't put everything into a single union because of conflicts between `T` in `Callable[..., T]` and other types, like generators and iterables
```py
import datetime
import typing
from typing import (
    Any,
    AsyncGenerator,
    AsyncIterable,
    Awaitable,
    Callable,
    Generator,
    Iterable,
    TypeVar,
    Union,
)

T = TypeVar("T")

_DependenciesWithYield = Union[
    Callable[..., Iterable[T]],
    Callable[..., Generator[T, None, None]],
    Callable[..., AsyncIterable[T]],
    Callable[..., AsyncGenerator[T, None]],
]


@typing.overload
def Depends(dependency: None = None) -> Any:
    ...


@typing.overload
def Depends(dependency: _DependenciesWithYield[T]) -> T:
    ...


@typing.overload
def Depends(dependency: Union[Callable[..., T], Callable[..., Awaitable[T]]]) -> T:
    ...


def Depends(dependency = None) -> Any:
    # Implementation
    pass


class Service:
    pass


def dependency_with_yield() -> Iterable[int]:
    yield int()


async def async_dependency_with_yield() -> AsyncIterable[int]:
    yield int()


service: Service = Depends()
reveal_type(service)  # Revealed type is "Service"
reveal_type(Depends(Service))  # Revealed type is "Service*"
reveal_type(Depends(int))  # Revealed type is "builtins.int*
reveal_type(Depends(dependency_with_yield))  # Revealed type is "builtins.int*
reveal_type(Depends(async_dependency_with_yield))  # Revealed type is "builtins.int*
reveal_type(Depends(datetime.datetime.now))  # Revealed type is "datetime.datetime*"
```

**Comment by odiseo0** [comment]
> I think something like this should do I had to create an overload for simple callables, seems like i can't put everything into a single union because of conflicts between `T` in `Callable[..., T]` and other types, like generators and iterables
> 
> ```python
> import datetime
> import typing
> from typing import (
>     Any,
>     AsyncGenerator,
>     AsyncIterable,
>     Awaitable,
>     Callable,
>     Generator,
>     Iterable,
>     TypeVar,
>     Union,
> )
> 
> T = TypeVar("T")
> 
> _DependenciesWithYield = Union[
>     Callable[..., Iterable[T]],
>     Callable[..., Generator[T, None, None]],
>     Callable[..., AsyncIterable[T]],
>     Callable[..., AsyncGenerator[T, None]],
> ]
> 
> 
> @typing.overload
> def Depends(dependency: None = None) -> Any:
>     ...
> 
> 
> @typing.overload
> def Depends(dependency: _DependenciesWithYield[T]) -> T:
>     ...
> 
> 
> @typing.overload
> def Depends(dependency: Union[Callable[..., T], Callable[..., Awaitable[T]]]) -> T:
>     ...
> 
> 
> def Depends(dependency = None) -> Any:
>     # Implementation
>     pass
> 
> 
> class Service:
>     pass
> 
> 
> def dependency_with_yield() -> Iterable[int]:
>     yield int()
> 
> 
> async def async_dependency_with_yield() -> AsyncIterable[int]:
>     yield int()
> 
> 
> service: Service = Depends()
> reveal_type(service)  # Revealed type is "Service"
> reveal_type(Depends(Service))  # Revealed type is "Service*"
> reveal_type(Depends(int))  # Revealed type is "builtins.int*
> reveal_type(Depends(dependency_with_yield))  # Revealed type is "builtins.int*
> reveal_type(Depends(async_dependency_with_yield))  # Revealed type is "builtins.int*
> reveal_type(Depends(datetime.datetime.now))  # Revealed type is "datetime.datetime*"
> ```

This looks fine to me 😁

**Comment by ThirVondukr** [comment]
After some analysis i think it's not possible to add correct type annotation to `Depends` mostly because of context manager-like dependencies that are usually annotated with `Iterable[T]` or `Generator[T, None, None]`

If we assume that all dependencies annotated with `Iterable[T]` return `T` would break dependencies that return iterable objects:
```py
_DependenciesWithYield = Union[
    Callable[..., Iterable[T]],
    Callable[..., Generator[T, None, None]],
    Callable[..., AsyncIterable[T]],
    Callable[..., AsyncGenerator[T, None]],
]


@typing.overload
def Depends(dependency: _DependenciesWithYield[T]) -> T:
    ...


def dependency() -> list[int]:
    return [42]
    

def dependency_with_yield() -> Iterable[T]:
    yield 42


reveal_type(Depends(dependency))  # int, Correct return type is `list[int]` or `Iterable[int]`
reveal_type(Depends(dependency_with_yield))  # int
```

Perhaps some context manager specific type annotations could be used to resolve problem with iterables and generators and wrap fastapi dependencies into `@contextlib.contextmanager` manually

**Comment by invokermain** [comment]
Is there not a middle ground where we can forward the return type if we are 100% sure, and otherwise return Any? It's a massive shame to discard this completely because of how useful it would be for FastAPI users? Something like:

```python
_DependenciesWithYield = Union[
    Callable[..., Iterable[T]],
    Callable[..., Generator[T, None, None]],
    Callable[..., AsyncIterable[T]],
    Callable[..., AsyncGenerator[T, None]],
]

@typing.overload
def Depends(dependency: None = None) -> Any:
    ...


@typing.overload
def Depends(dependency: _DependenciesWithYield[T]) -> Any:
    ...


@typing.overload
def Depends(dependency: Union[Callable[..., T], Callable[..., Awaitable[T]]]) -> T:
    ...


def Depends(dependency = None) -> Any:
    # Implementation
    pass
```

**Comment by ThirVondukr** [comment]
@invokermain Maybe, we also don't need `Callable[..., Awaitable[T]]` annotation, just `Callable[..., T]`
